proofs.go

```diff
@@ -119,37 +126,154 @@ func (proof *SparseCompactMerkleProof) Unmarshal(bz []byte) error {
+	if proof.Depth < 0 || proof.Depth > spec.ph.PathSize()*8 {
+		return fmt.Errorf("invalid depth: %d", proof.Depth)
 	}
+	// for each of the bits flipped ensure that they are within the path size
+	// and that they are not greater than the depth of the leaf node being proven
+	for _, i := range proof.FlippedBits {
+		// as proof.Depth <= spec.ph.PathSize()*8, i <= proof.Depth
+		if i < 0 || i > proof.Depth {
+			return fmt.Errorf("invalid flipped bit index: %d", i)
+		}
+	}
+	// create the path of the leaf node using the flipped bits metadata
+	workingPath := proof.Path
+	for _, i := range proof.FlippedBits {
+		flipPathBit(workingPath, i)
+	}
+	// ensure that the path of the leaf node being proven has a prefix
+	// of length depth as the path provided (with bits flipped)
+	if equal, failed := equalPrefixBits(
+		workingPath,
+		proof.ClosestPath,
+		0, proof.Depth,
+	); !equal {
+		return fmt.Errorf("invalid closest path: %x (not equal at bit: %d)", proof.ClosestPath, failed)
+	}
+	// validate the proof itself
+	if err := proof.ClosestProof.validateBasic(spec); err != nil {
+		return fmt.Errorf("invalid closest proof: %w", err)
+	}
+	return nil
+}
 
-	return true
+// SparseCompactMerkleClosestProof is a compressed representation of the SparseMerkleClosestProof
+type SparseCompactMerkleClosestProof struct {
+	Path             []byte                    // the path provided to the ProveClosest method
+	FlippedBits      [][]byte                  // the index of the bits flipped in the path during tree traversal
+	Depth            []byte                    // the depth of the tree when tree traversal stopped
+	ClosestPath      []byte                    // the path of the leaf closest to the path provided
+	ClosestValueHash []byte                    // the value hash of the leaf closest to the path provided
+	ClosestProof     *SparseCompactMerkleProof // the proof of the leaf closest to the path provided
+}
+
+func (proof *SparseCompactMerkleClosestProof) validateBasic(spec *TreeSpec) error {
+	// Do a basic sanity check on the proof on the fields of the proof specific to
+	// the compact proof only.
+	//
+	// When the proof is de-compacted and verified, the sanity check for the
+	// de-compacted proof should be executed.
+
+	// ensure no compressed fields are larger than the path size
+	maxSliceLen := minBytes(spec.ph.PathSize() * 8)
+	if len(proof.Depth) > maxSliceLen {
+		return fmt.Errorf("invalid depth: %d", proof.Depth)
+	}
+	for _, i := range proof.FlippedBits {
+		if len(i) > maxSliceLen {
+			return fmt.Errorf("invalid flipped bit index: %d", bytesToInt(i))
+		}
+	}
+	// perform a sanity check on the closest proof
+	if err := proof.ClosestProof.validateBasic(spec); err != nil {
+		return fmt.Errorf("invalid closest proof: %w", err)
+	}
+	return nil
+}
+
+// Marshal serialises the SparseCompactMerkleClosestProof to bytes
+func (proof *SparseCompactMerkleClosestProof) Marshal() ([]byte, error) {
```

Comment:
```suggestion
// TODO(#22): Use protos instead of bytes for this in the future
func (proof *SparseCompactMerkleClosestProof) Marshal() ([]byte, error) {
```

Reply:
Added at the top